### DISCUSSION TITLE
How to config `nvim_lspconfig` to work with `markdown` & `quarto` files using `lazyvim` as template?

### DISCUSSION THREAD

**Discussion by hopezh** [quarto-dev/quarto-nvim#82] [Q&A] [answered]
# 1. Objective
I'm using [lazyvim](https://www.lazyvim.org/) as a template to create my own `neovim` config [oatLazyVim](https://github.com/hopezh/oatLazyVim), and I want to make it work for `quarto` `.qmd` files to enjoy all the `lsp` functions for `markdown` and `python` codes.

# 2. What I tried
I mentioned in [a previous discussion](https://github.com/quarto-dev/quarto-nvim/discussions/75#discussioncomment-6352000) that I've managed "to make marksman work for both markdown md and quarto qmd files. Both [symbols-outline.nvim](https://github.com/simrat39/symbols-outline.nvim) and [nvim-navbuddy](https://github.com/SmiteshP/nvim-navbuddy) can recognize the headings in qmd files without including the comment symbol # in a code block", as shown by the code below.  

```Lua
return {
    -- add pyright and marksman to lspconfig
    {
        "neovim/nvim-lspconfig",

        ---@class PluginLspOpts
        opts = {
            ---@type lspconfig.options
            servers = {
                -- pyright will be automatically installed with mason and loaded with lspconfig
                pyright = {},
                marksman = {},
            },
        },

        -- add marksman support for markdown and quoarto files
        config = function()
            local lspconfig = require("lspconfig")
            local cmp_nvim_lsp = require("cmp_nvim_lsp")
            local util = require("lspconfig.util")

            local on_attach2 = function(client, bufnr)
                local function buf_set_keymap(...)
                    vim.api.nvim_buf_set_keymap(bufnr, ...)
                end
                local function buf_set_option(...)
                    vim.api.nvim_buf_set_option(bufnr, ...)
                end

                buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")
                local opts = { noremap = true, silent = true }

                buf_set_keymap("n", "gh", "<cmd>lua vim.lsp.buf.signature_help()<CR>", opts)
                buf_set_keymap("n", "gi", "<cmd>Telescope lsp_implementations<CR>", opts)
                buf_set_keymap("n", "[d", "<cmd>lua vim.diagnostic.goto_prev()<CR>", opts)
                buf_set_keymap("n", "]d", "<cmd>lua vim.diagnostic.goto_next()<CR>", opts)
                buf_set_keymap("n", "<leader>ll", "<cmd>lua vim.lsp.codelens.run()<cr>", opts)
                client.server_capabilities.document_formatting = true
            end

            local capabilities = vim.lsp.protocol.make_client_capabilities()
            capabilities = cmp_nvim_lsp.default_capabilities(capabilities)
            capabilities.textDocument.completion.completionItem.snippetSupport = true

            -- also need to add the following to $home/.config/marksman/config.toml :
            -- [core]
            -- markdown.file_extensions = ["md", "markdown", "qmd"]
            lspconfig.marksman.setup({
                on_attach = on_attach2,
                capabilities = capabilities,
                filetypes = { "markdown", "quarto" },
                root_dir = util.root_pattern(".git", ".marksman.toml", "_quarto.yml"),
            })
        end,
    },
}
```

# 3. The problem
However, this config seemed to interfere with and wash away some of the default `nvim-lspconfig` settings came with `lazyvim`, especially some of the keybindings.  

# 4. My question
I want to ask how to add configs of `marksman` for `markdown` and `quarto` files in `nvim-lspconfig` in "lazyvim's way", without affecting the default settings of `nvim-lspconfig` came with `lazyvim` installation. 

More specifically, I noticed the `marksman` related settings as shown in the [quarto.lua](https://github.com/jmbuhr/quarto-nvim-kickstarter/blob/main/lua/plugins/quarto.lua) file in [quarto-nvim-kickstarter](https://github.com/jmbuhr/quarto-nvim-kickstarter), such as the following block of codes that involves the use of variables such as `on_attach` and `capabilities`: 

```Lua
      lspconfig.marksman.setup {
        on_attach = on_attach_qmd,
        capabilities = capabilities,
        filetypes = { 'markdown', 'quarto' },
        root_dir = util.root_pattern(".git", ".marksman.toml", "_quarto.yml"),
      }
```

So, how to achieve the same in the `marksman` related section in the [lua/plugins/nvim_lspconfig.lua](https://github.com/hopezh/oatLazyVim/blob/main/lua/plugins/nvim_lspconfig.lua) file in a `lazyvim` config, as shown below, without messing up the default settings and keymaps of `nvim-lspconfig`? 

```Lua
return {
    -- add pyright and marksman to lspconfig
    {
        "neovim/nvim-lspconfig",

        ---@class PluginLspOpts
        opts = {
            ---@type lspconfig.options
            servers = {
                -- pyright will be automatically installed with mason and loaded with lspconfig
                pyright = {},
                marksman = {
                    filetypes = { "markdown", "quarto" },
                },
            },
        },
    },
}
```

**Comment by jmbuhr** [comment, marked as answer]
I made a version of kickstarter based on LazyVim for this very purpose. 
https://github.com/jmbuhr/lazyvim-starter-for-quarto/blob/main/lua/plugins/quarto.lua
should contain everything you need.

Incidentally, the on_attach things are not necessary there because LazyVim uses the LspAttach autocommand, which runs before on_attach, so I was able to add my own autocommand right inside of quarto/otter (here: https://github.com/quarto-dev/quarto-nvim/blob/main/ftplugin/quarto.lua). If quarto is configured after lspconfig (determined via the dependencies), it's keybindings should win out. This should be the case in the example.

**Comment by hopezh** [comment]
@jmbuhr Thank you very much. The [lazyvim-starter-for-quarto](https://github.com/jmbuhr/lazyvim-starter-for-quarto/tree/main) is very helpful. 

I modifed my `oatLazyVim` config accordingly in a new [branch](https://github.com/hopezh/oatLazyVim/tree/test-lazyvim-starter-for-quarto) for test. 

I didn't add any keymap for `quarto-nvim`, like those specified in your [quarto.lua](https://github.com/jmbuhr/lazyvim-starter-for-quarto/blob/4535fc121915b1e70822fc940760ee4e2f671832/lua/plugins/quarto.lua#L21-L39) file in the [lazyvim-starter-for-quarto](https://github.com/jmbuhr/lazyvim-starter-for-quarto/tree/main), except for using `<leader>qp` for `Quarto Preview`, in my [/lua/plugins/quarto.lua](https://github.com/hopezh/oatLazyVim/blob/test-lazyvim-starter-for-quarto/lua/plugins/quarto.lua) : 

```Lua
    {
        "quarto-dev/quarto-nvim",

        dev = false,
        ft = "quarto",

        -- dependencies = {
        --     "neovim/nvim-lspconfig",
        -- },

        opts = {
            lspFeatures = {
                languages = { "r", "python", "julia", "bash", "lua", "html" },
            },
        },

        keys = {
            {
                "<leader>qp",
                "<CMD>QuartoPreview<CR>",
                desc = "Quarto Preview",
            },
        },
    },
```

However, when I press the `<leader>` key, the `which-key` prompt is different from that of `lazyvim's` default when editing a `qmd` file. 

Firstly, the `l` key is changed to `prefix`, rather than for `Lazy`. 

`which-key` prompt when editing `qmd` file:
<img width="1269" alt="Screenshot 2023-07-30 at 5 00 55 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/503af512-28ba-43f4-9cf4-f0ca0f14d351">

`which-key` prompt when edting a `python` file: 
<img width="1269" alt="Screenshot 2023-07-30 at 5 01 16 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/9f54511a-07ec-4f2c-b07e-0cd8a16260cd">

Secondly, the keymap for `rename` is changed to `<leader>lR` for `require'otter'.ask_rename()`, and `lazyvim's` keymap `<leader>cr` is not working anymore:
<img width="1269" alt="Screenshot 2023-07-30 at 5 02 21 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/d6ecdf07-2ea6-4fcc-995e-8836ff07c856">

I want to ask:  
Where and when is the keymap `<leader>lR` for `require'otter'.ask_rename()` defined, because I didn't add it in [my config](https://github.com/hopezh/oatLazyVim/blob/test-lazyvim-starter-for-quarto/lua/plugins/quarto.lua), and I didn't add the [/quarto-nvim/ftplugin/quarto.lua](https://github.com/quarto-dev/quarto-nvim/blob/main/ftplugin/quarto.lua) file, either?

**Reply by jmbuhr** [reply]
It is set here: https://github.com/quarto-dev/quarto-nvim/blob/main/ftplugin/quarto.lua and comes from the quarto.nvim config (https://github.com/quarto-dev/quarto-nvim/blob/35f86035e7b3846dbf168267ffe0021c3d312259/lua/quarto/init.lua#L8-L34), so you can change the keybinding on setup.

**Reply by hopezh** [reply]
Thanks, @jmbuhr 

I've edited [my config](https://github.com/hopezh/oatLazyVim/tree/test-lazyvim-starter-for-quarto) and the [/lua/plugins/quarto.lua](https://github.com/hopezh/oatLazyVim/blob/test-lazyvim-starter-for-quarto/lua/plugins/quarto.lua) in which I specified `<leader>cr` for `rename` in the `keymap` section of `opts`:

```Lua
    {
        "quarto-dev/quarto-nvim",

        dev = false,
        ft = "quarto",

        dependencies = {
            "neovim/nvim-lspconfig",
            "jmbuhr/otter.nvim",
            "hrsh7th/nvim-cmp",
        },

        opts = {
            lspFeatures = {
                languages = { "r", "python", "julia", "bash", "lua", "html" },
            },
            keymap = {
                -- rename = '<leader>lR', -- default of quarto-nvim
                rename = "<leader>cr",
            },
        },

        keys = {
            {
                "<leader>qp",
                "<CMD>QuartoPreview<CR>",
                desc = "Quarto Preview",
            },
        },
    },
```

Now, `<leader>cr` is shown as `r -> require 'otter'.ask_rename()` in `which-key` prompt:
<img width="1272" alt="Screenshot 2023-07-30 at 8 42 39 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/2caccc5d-dcdb-4a99-8ac7-70ad5a5d529f">

However, when I applied this keymap to rename the variable `r` in the `python` code in a [hello.qmd](https://github.com/hopezh/oatLazyVim/blob/test-lazyvim-starter-for-quarto/test/hello.qmd) test file I got the following error, and I was unable to rename the variable. 

```Shell
Error executing vim.schedule lua callback: ...ocal/share/OatLazyVim/lazy/otter.nvim/lua/otter/init.lua:185: bad argument #1 to 'pairs' (table expected, got nil)
stack traceback:
	[C]: in function 'pairs'
	...ocal/share/OatLazyVim/lazy/otter.nvim/lua/otter/init.lua:185: in function 'filter'
	...al/share/OatLazyVim/lazy/otter.nvim/lua/otter/keeper.lua:340: in function 'handler'
	...r/neovim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp.lua:1499: in function 'cb'
	vim/_editor.lua:324: in function <vim/_editor.lua:323>
```
<img width="2552" alt="Screenshot 2023-07-30 at 8 43 32 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/49938de9-c483-4b02-940b-2f0c1ccefa9c">

Seems it's related to `otter` settings, but I don't know the exact where it went wrong... 

Can you kindly advise what I might have missed here that leads to this error? 

Thanks.

**Reply by jmbuhr** [reply]
You haven't missed anything, that's actually a bug in otter.nvim that I haven't managed to pin down exactly, yet. I think something changed about the way I'm supposed to pass the current variable name to the prompt that asks for the new name in neovim or lspconfig.

**Reply by hopezh** [reply]
Hi @jmbuhr, may I ask if the error related to renaming a python variable in `qmd` file as reported above has been resolved?

**Reply by jmbuhr** [reply]
yes, It should be fixed in https://github.com/jmbuhr/otter.nvim/releases/tag/v1.0.2
Do let me know if you find a counter example.

**Comment by hopezh** [comment]
> yes, It should be fixed in https://github.com/jmbuhr/otter.nvim/releases/tag/v1.0.2 Do let me know if you find a counter example.

Hi @jmbuhr, thank you. 

However, after I updated quarto and otter plugins in my config, I'm getting the following error while trying to rename a variable in python:

```
Error executing vim.schedule lua callback: ...vim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/util.lua:496: bad argument #1 to 'gsub' (string expected, got nil)
stack traceback:
	[C]: in function 'gsub'
	...vim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/util.lua:496: in function 'apply_text_edits'
	...vim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/util.lua:655: in function 'apply_text_document_edit'
	...vim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/util.lua:916: in function 'apply_workspace_edit'
	...HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/handlers.lua:305: in function <...HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp/handlers.lua:299>
	...al/share/OatLazyVim/lazy/otter.nvim/lua/otter/keeper.lua:346: in function 'handler'
	...r/neovim/HEAD-e59cf3b/share/nvim/runtime/lua/vim/lsp.lua:1499: in function 'cb'
	vim/_editor.lua:324: in function <vim/_editor.lua:323>
```

<img width="1556" alt="Screenshot 2023-08-18 at 10 36 01 PM" src="https://github.com/quarto-dev/quarto-nvim/assets/23435810/0b9e4d41-f9b2-48e3-aa03-015086fe1fa9">

My [quarto.lua](https://github.com/hopezh/oatLazyVim/blob/main/lua/plugins/quarto.lua) file is in this [oatLazyVim](https://github.com/hopezh/oatLazyVim) config. 

I used this [hello.qmd](https://github.com/hopezh/oatLazyVim/blob/main/test/hello.qmd) file to test renaming the variable "r" in the python code block.

**Reply by jmbuhr** [reply]
are you in a python/git project? what version of nvim are you on?

**Reply by hopezh** [reply]
My nvim version is:
```
NVIM v0.10.0-dev-575+ge59cf3b3a-Homebrew
Build type: Release
LuaJIT 2.1.0-beta3
```

I have no problem using lazyvim's default keybinding `<leader>cr` for rename when editing a pure python script. 

I only have such issue with "rename" when editing python code block in a qmd file, using the same "<leader>cr" keybinding, and the which-key prompt for it changed to "r -> require'otter'.ask_rename()".

**Reply by jmbuhr** [reply]
I your quarto document part of a python project?

**Reply by jmbuhr** [reply]
Found the problem. The LazyVim distro has a plugin enabled by default that overwrites the vim.ui.input method (to place it next to the cursor instead of in the command line), but that doesn't play nicely with otter.nvim (something is happening asynchronously that shouldn't be, I believe).
For now I fixed it by simply disabling the offending plugin: https://github.com/jmbuhr/lazyvim-starter-for-quarto/commit/79cbe55ff5c094db3577f1baad603aa0ab231a8c#diff-ae0a74392a85a01dcf65a8b24a487b6035159ab5a5801df01320e45706ecd32c

**Reply by jmbuhr** [reply]
Thanks for catching this bug :)

**Reply by hopezh** [reply]
Yes, disabling `dressing.nvim` works like a charm.
 
Thank you @jmbuhr for the prompt fix.
